ui/apps/platform/src/Containers/Policies/Wizard/Step3/PolicyCriteriaOptions.tsx

```diff
@@ -0,0 +1,188 @@
+    const categories = getKeysByCategory(descriptors);
+
+    const treeList = Object.keys(categories).map((category) => ({
+        name: '',
+        title: category,
+        id: kebabCase(category),
+        children: categories[category]
+            .filter((child) => {
+                const alreadyUsed = existingGroups.find(
+                    (group) =>
+                        group.fieldName.toLowerCase() === child?.name?.toLowerCase() ||
+                        group.fieldName.toLowerCase() === child?.label?.toLowerCase()
+                );
+                return !alreadyUsed;
+            })
+            .map<TreeViewDataItem>((child: Descriptor) => ({
+                name: child.longName,
+                title: child.shortName || child.name,
+                id: kebabCase(child.shortName),
+                icon: <PlusCircleIcon />,
+            })),
+    }));
+
+    return treeList;
+}
+
+type PolicyCriteriaOptionsProps = {
+    descriptors: Descriptor[];
+    selectedSectionIndex: number;
+};
+
+function PolicyCriteriaOptions({ descriptors, selectedSectionIndex }: PolicyCriteriaOptionsProps) {
+    // const [activeItems, setActiveItems] = useState<TreeViewDataItem[]>([]);
+    const [allExpanded, setAllExpanded] = useState(false);
+    const [filteredItems, setFilteredItems] = useState<TreeViewDataItem[]>([]);
+    const [isFiltered, setIsFiltered] = useState(false);
+    const { values, setFieldValue } = useFormikContext<Policy>();
+    const { policyGroups } = values.policySections[selectedSectionIndex] ?? [];
+
+    const selectedSection = values.policySections[selectedSectionIndex];
+
+    const treeDataItems = getPolicyFieldsAsTree([], descriptors);
+
+    useEffect(() => {
+        setFilteredItems(treeDataItems);
+        setIsFiltered(false);
+    }, [treeDataItems]);
+
+    function onSearch(evt) {
+        const input: string = evt.target.value;
+        if (input === '') {
+            setFilteredItems(treeDataItems);
+            setIsFiltered(false);
+        } else {
+            const filtered = treeDataItems.map((item) => {
+                const filteredItem = { ...item };
+                if (item.children && item.children.length > 0) {
+                    const filteredChildren = item.children.filter((child) => {
+                        const name = typeof child.name === 'string' ? child.name : '';
+                        const title = typeof child.title === 'string' ? child.title : '';
+                        return (
+                            name.toLowerCase().includes(input.toLowerCase()) ||
+                            title.toLowerCase().includes(input.toLowerCase())
+                        );
+                    });
+
+                    filteredItem.children = filteredChildren;
+                }
+
+                return filteredItem;
+            });
+
+            setFilteredItems(filtered);
+            setIsFiltered(true);
+        }
+    }
```

Comment:
Instead of lifting the state up like we were discussing in another comment, would having local state for the search filter here clear up the problems with `useEffect`?

We could do something like:
```ts
const [searchInput, setSearchInput] = useState();
const treeDataItems = getPolicyFieldsAsTree([], descriptors);
const isFiltered = searchInput !== '';
const filteredItems = isFiltered ? fnThatFilters(treeDataItems) : treeDataItems;

function onSearch() {
    // set input value only
}
```

This would do some level of calculation during render, but if it looks like be a performance problem we could `useMemo` against the value of the searchInput to cache it. (So long as the `descriptors` passed in do not change, otherwise we might need to memo those as well.)